sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1764,6 +1765,58 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns a new Dataset containing union of rows in this Dataset and another Dataset.
+   *
+   * This is different from both `UNION ALL` and `UNION DISTINCT` in SQL. To do a SQL-style set
+   * union (that does deduplication of elements), use this function followed by a [[distinct]].
+   *
+   * The difference between this function and [[union]] is that this function
+   * resolves columns by name (not by position):
+   *
+   * {{{
+   *   val df1 = Seq((1, 2, 3)).toDF("col0", "col1", "col2")
+   *   val df2 = Seq((4, 5, 6)).toDF("col1", "col2", "col0")
+   *   df1.unionByName(df2).show
+   *
+   *   // output:
+   *   // +----+----+----+
+   *   // |col0|col1|col2|
+   *   // +----+----+----+
+   *   // |   1|   2|   3|
+   *   // |   6|   4|   5|
+   *   // +----+----+----+
+   * }}}
+   *
+   * @group typedrel
+   * @since 2.3.0
+   */
+  def unionByName(other: Dataset[T]): Dataset[T] = withSetOperator {
+    // Creates a `Union` node and resolves it first to reorder output attributes in `other` by name
+    val unionPlan = sparkSession.sessionState.executePlan(Union(logicalPlan, other.logicalPlan))
+    unionPlan.assertAnalyzed()
+    val Seq(left, right) = unionPlan.analyzed.children
+
+    // Builds a project list for `other` based on `logicalPlan` output names
+    val resolver = sparkSession.sessionState.analyzer.resolver
+    val rightProjectList = mutable.ArrayBuffer.empty[Attribute]
+    val rightOutputAttrs = right.output
+    for (lattr <- left.output) {
+      // To handle duplicate names, we first compute diff between `rightOutputAttrs` and
+      // already-found attrs in `rightProjectList`.
+      rightOutputAttrs.diff(rightProjectList).find { rattr => resolver(lattr.name, rattr.name)}
```

Comment:
In the logic, it seems we cannot catch left column duplication, I think.
How about checking column name duplication first, then build a right project list?
```

    // Check column name duplication in both sides first
    val leftOutputAttrs = left.output
    val rightOutputAttrs = right.output
    val caseSensitiveAnalysis = sparkSession.sessionState.conf.caseSensitiveAnalysis
    SchemaUtils.checkColumnNameDuplication(
      leftOutputAttrs.map(_.name), "left column names", caseSensitiveAnalysis)
    SchemaUtils.checkColumnNameDuplication(
      rightOutputAttrs.map(_.name), "right column names", caseSensitiveAnalysis)

    // Then, builds a project list for `other` based on `logicalPlan` output names
    val resolver = sparkSession.sessionState.analyzer.resolver
    val rightProjectList = left.output.map { lattr =>
      val foundAttrs = rightOutputAttrs.filter { rattr => resolver(lattr.name, rattr.name) }
      assert(foundAttrs.size > 1)
      if (foundAttrs.size == 1) {
        foundAttrs.head
      } else if (foundAttrs.size == 0) {
        throw new AnalysisException(s"""Cannot resolve column name "${lattr.name}" among """ +
          s"""(${rightOutputAttrs.map(_.name).mkString(", ")})""")
      }
    }
```

(I used `SchemaUtils` here implemented in #17758 
https://github.com/apache/spark/pull/17758/files#diff-dc9b15e4af298799d788b59d2baf96a9R29)